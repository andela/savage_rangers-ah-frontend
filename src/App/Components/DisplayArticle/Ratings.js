/* eslint-disable no-restricted-globals */
import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { ToastContainer, toast } from 'react-toastify';
import propTypes from 'prop-types';
import isEmpty from 'lodash/isEmpty';
import Rater from 'react-star-ratings';
import Rating from './Rating';
import actions from '../../../Redux/Actions/articleRatings';

const { getRatings, rate } = actions;

export class Ratings extends Component {
  static propTypes = {
    getRatings: propTypes.func,
    ratings: propTypes.object,
    errorMessage: propTypes.string,
    successMessage: propTypes.string,
    rate: propTypes.func,
    articleSlug: propTypes.string.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      average: null,
      allUsers: null,
      stats: [],
      rating: 0,
      redirect: false,
      isLoggedIn: false,
      defaultError: ''
    };
  }

  componentDidMount() {
    const { props: { articleSlug } } = this;
    const { getRatings: getArticleRatings } = this.props;
    getArticleRatings(articleSlug);
  }

  componentWillReceiveProps({ ratings, errorMessage, successMessage }) {
    const { state: { defaultError } } = this;
    if (!isEmpty(ratings)) {
      this.setState({
        average: ratings.average.toString().substring(0, 3),
        allUsers: ratings.data.allUsers,
        stats: ratings.data.statistics
      });
    }

    if (
      errorMessage
      && errorMessage !== defaultError
      && errorMessage !== 'No ratings found for this article'
    ) {
      this.setState({ defaultError: errorMessage });
      toast.error(errorMessage);
    }
    if (successMessage) toast.success(successMessage);
  }

  rate = async (rating) => {
    const redirectToLogin = () => this.setState({ redirect: true });
    const Toast = ({ closeToast }) => (
      <div>
        <p className="ogin-popup-message">Please, login first...</p>
        <p className="login-popup-buttons">
          <button type="submit" onClick={redirectToLogin}>
            Yes
          </button>
          <button type="submit" onClick={closeToast}>
            No
          </button>
        </p>
      </div>
    );

    const token = localStorage.getItem('token');
    const { props: { articleSlug } } = this;
    if (token) {
      this.setState({ isLoggedIn: true });
      const {
        props: { rate: postRating, getRatings: getArticleRatings },
        state: { rating: userRating, defaultError }
      } = this;
      if (rating !== userRating) {
        this.setState({ rating });
        postRating(rating, articleSlug, token).then(() => {
          getArticleRatings(articleSlug);
        });
      } else toast.error(defaultError);
    } else {
      toast.error(Toast);
    }
  };

  render() {
    const {
      state: {
        stats, average, allUsers, redirect, isLoggedIn
      }
    } = this;
    // Ordering the array of stats
    const orderedStats = [1, 2, 3, 4, 5];
    stats.map((item) => {
      orderedStats[item.rating - 1] = item;
      return item;
    });
    if (redirect) return <Redirect to={`/login?redirect=${location.pathname}`} />;
    return (
      <React.Fragment>
        {!isLoggedIn && <ToastContainer autoClose={false} closeButton={false} />}
        {!isEmpty(stats) ? (
          <div className="ratings-container">
            <div className="ratings-container--stats">
              <p className="ratings-container--stats__average">{average}</p>
              <p className="ratings-container--stats__viewers">
                {allUsers}
                {' '}
                {allUsers > 1 ? 'Reviews' : 'Review'}
              </p>
              <div className="custom-rater">
                <Rater
                  numOfStars={5}
                  rating={parseFloat(average)}
                  starRatedColor="#ffd700"
                  starHoverColor="#d4b50f"
                  name="rating"
                  starDimension="20px"
                  starSpacing="15px"
                  changeRating={this.rate}
                />
              </div>
            </div>
            {orderedStats.reverse().map(({ rating, percentage }) => (
              <Rating key={rating} rating={rating} percentage={percentage} stats={stats} />
            ))}
          </div>
        ) : (
          <div className="custom-rater-default">
            <p>Rate this article</p>
            <Rater
              numOfStars={5}
              rating={0}
              starRatedColor="#ffd700"
              starHoverColor="#d4b50f"
              name="rating"
              starDimension="20px"
              starSpacing="10px"
              changeRating={this.rate}
            />
          </div>
        )}
      </React.Fragment>
    );
  }
}

export const mapStateToProps = ({ articleRatings: { data, errorMessage, successMessage } }) => ({
  ratings: data,
  errorMessage,
  successMessage
});

export default connect(mapStateToProps,
  { getRatings, rate })(Ratings);
