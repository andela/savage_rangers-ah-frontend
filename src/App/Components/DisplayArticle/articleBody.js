import React from 'react';
import ReactImageFallback from 'react-image-fallback';
import ReactHtmlParser from 'react-html-parser';
import moment from 'moment';
import Tags from './tags';

const ArticleBody = (props) => {
  const {
    title, body,
    readTime, createdAt,
    coverImage, firstName,
    lastName, profileImage, tags
  } = props;
  const formattedDate = moment(createdAt).format('Do MMM YYYY');
  return (
    <div className="col-xl-9 col-lg-9 col-md-12 col-sm-12 col-12">
      <div className="article__data">
        <h2 className="article__description">{title}</h2>
        <div className="article__profileDetails">
          <ReactImageFallback className="article__imgProfile" src={profileImage} fallbackImage="https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png" alt="profilePic" title="profile" />
          <p to="/" className="article__nameDetails">{`${firstName} ${lastName}`}</p>
          <p to="/" className="article__dateDetails">{formattedDate}</p>
          <div className="article__point-separator"> . </div>
          <p to="/" className="article__minDetails">
            {readTime}
            {' '}
            min read
          </p>
        </div>
        <div className="DraftEditor-root">
          <div className="DraftEditor-editorContainer">
            <div className="ml-0.5">{<ReactImageFallback className="DraftEditor-coverImage ml-1" src={coverImage} fallbackImage="https://ielektro.es/wp-content/uploads/2017/04/ventajas-comprar-LED.jpg" alt="coverImage" />}</div>
            <div className="public-DraftEditor-content">
              {ReactHtmlParser(body)}
            </div>
          </div>
        </div>
        <Tags tags={tags} />
      </div>
    </div>
  );
};

export default ArticleBody;
