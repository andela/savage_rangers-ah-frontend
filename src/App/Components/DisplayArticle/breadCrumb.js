import React from 'react';

const BreadCrumb = (props) => {
  const { category } = props;
  return (
    <div className="page__label col-lg-2" id="breadcrumb">
      <p className="page__label--text">Articles</p>
      <div className="category-p">
        <p className="category-p__label">CATEGORY</p>
      </div>
      <div className="category-t">
        <p className="category-t__label">{category}</p>
      </div>
    </div>
  );
};


export default BreadCrumb;
