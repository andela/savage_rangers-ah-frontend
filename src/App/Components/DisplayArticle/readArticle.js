/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
// import Popular from '../Popular/popular';
import ReadArticleActions from '../../../Redux/Actions/readArticleActions';
import PopularArticleAction from '../../../Redux/Actions/readPopularActions';
import ArticleNotFound from '../ArticleNotFound/ArticleNotFound';
import BreadCrumb from './breadCrumb';
import ArticleBody from './articleBody';
import Loader from '../Common/loader';
import Navbar from '../Common/NavProfile/navbar';
import Footer from '../Common/Footer';
import BottomPopular from '../Popular/bottom-popular';


const { readArticle, getTags } = ReadArticleActions;
const { readPopularArticle } = PopularArticleAction;

export class ReadArticle extends Component {
  constructor(props) {
    super(props);
    this.state = { tags: [], isLoading: true, articles: [{ title: '', User: {}, Category: {} }] };
  }


  componentDidMount() {
    const {
      readArticle: readArticles,
      getTags: getTag,
      readPopularArticle: readPopular
    } = this.props;
    const { match: { params: { slug } } } = this.props;
    readArticles(slug);
    getTag(slug);
    readPopular();
  }


  componentWillReceiveProps(nextProps) {
    if (nextProps.article) {
      this.setState({
        slug: nextProps.article.slug,
        category: nextProps.article.Category.name,
        title: nextProps.article.title,
        body: nextProps.article.body,
        readTime: nextProps.article.readTime,
        createdAt: nextProps.article.createdAt,
        coverImage: nextProps.article.coverImage,
        firstName: nextProps.article.User.firstName,
        lastName: nextProps.article.User.lastName,
        profileImage: nextProps.article.User.profileImage
      });
    }

    setTimeout(() => {
      this.setState({ isLoading: false });
    }, 2000);

    if (nextProps.tags) {
      this.setState({ tags: nextProps.tags.data });
    }

    if (nextProps.popularArticle) {
      this.setState({ articles: nextProps.popularArticle.data });
    }

    const { readArticle: readArticles, getTags: getTag } = this.props;
    const { params: { slug } } = nextProps.match;

    if (nextProps.match.params.slug !== this.state.slug && this.state.slug) {
      this.setState({ isLoading: true });
      readArticles(slug);
      getTag(slug);
    }
  }

  render() {
    const {
      category, title, body, readTime, tags,
      createdAt, coverImage,
      firstName, lastName, profileImage, slug, isLoading, articles
    } = this.state;

    return isLoading ? <Loader /> : (
      <React.Fragment>
        {
          !slug ? <ArticleNotFound content="ARTICLE NOT FOUND" /> : (
            <div>
              <Navbar />
              <BreadCrumb category={category} />
              <div className="container article">
                <div className="row">
                  <ArticleBody
                    title={title}
                    body={body}
                    readTime={readTime}
                    tags={tags}
                    createdAt={createdAt}
                    coverImage={coverImage}
                    firstName={firstName}
                    lastName={lastName}
                    profileImage={profileImage}
                  />


                  {/* <div className="col-xl-2 col-lg-2 col-md-12 col-sm-12 col-12">
                    <Popular articles={articles} />
                  </div> */}

                  <BottomPopular articles={articles} />
                </div>
              </div>
              <Footer />
            </div>
          )
        }
      </React.Fragment>
    );
  }
}
ReadArticle.propTypes = { article: propTypes.object, tags: propTypes.object };
export const mapStateToProps = state => ({
  article: state.readArticle.article,
  tags: state.readArticle.tags,
  popularArticle: state.populars.Articles
});
export default connect(mapStateToProps, { readArticle, getTags, readPopularArticle })(ReadArticle);
