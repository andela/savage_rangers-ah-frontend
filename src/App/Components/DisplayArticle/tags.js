import React from 'react';

const Tags = (props) => {
  const { tags } = props;
  return (
    <div className="ml-2">
      {
        tags.map(item => (
          <div key={item.id} className="article__tags">
            {item.name}
          </div>
        ))
                  }

    </div>
  );
};

export default Tags;
