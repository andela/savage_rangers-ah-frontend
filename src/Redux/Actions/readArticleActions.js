import types from './index';
import axios from '../../configs/axios';

const { FETCH_ONE_ARTICLE, GET_ARTICLE_TAGS, READ_ARTICLE_ERROR } = types;

export default {
  readArticle: slug => (dispatch) => {
    axios
      .get(`/api/articles/${slug}`)
      .then((res) => {
        dispatch({
          type: FETCH_ONE_ARTICLE,
          payload: res.data.article
        });
      }).catch((err) => {
        const error = err.response.data.errors.Article;
        dispatch({ type: READ_ARTICLE_ERROR, payload: error });
      });
  },
  getTags: slug => (dispatch) => {
    axios
      .get(`/api/articles/${slug}/tags`)
      .then((res) => {
        dispatch({
          type: GET_ARTICLE_TAGS,
          payload: res.data
        });
      }).catch((err) => {
        const error = err.response.data.errors.tags;
        dispatch({ type: READ_ARTICLE_ERROR, payload: error });
      });
  }
};
