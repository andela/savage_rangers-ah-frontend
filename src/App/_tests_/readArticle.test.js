import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import mockAxios from 'axios';
import promiseMiddleware from 'redux-promise-middleware';
import { shallow } from '../../enzyme';
import readArticleActions from '../../Redux/Actions/readArticleActions';

import { ReadArticle, mapStateToProps } from '../Components/DisplayArticle/readArticle';
import Readreducer from '../../Redux/Reducers/readArticleReducer';

const { readArticle, getTags } = readArticleActions;
const middlewares = [thunk, promiseMiddleware];
const mockStore = configureStore(middlewares);

const initialState = { Article: {} };
const store = mockStore(initialState);

const payload = {
  article: {
    id: 6,
    title: 'What you need to know to become a great software engineer in 2020',
    description: '10 Tips to become a great software engineer by 2020.',
    body: 'I get quite a few emails that basically say',
    slug: 'what-you-need-to-know-to-become-a-great-software-engineer-in-2020-nfovo9fghng',
    readTime: 6,
    coverImage: 'https://res.cloudinary.com/al-tech/image/upload/v1565787533/rvkmbae3pcqhgysxoqnk.jpg',
    author: 42,
    category: 4,
    isBlocked: false,
    status: 'published',
    createdAt: '2019-08-14T12:58:54.185Z',
    updatedAt: '2019-08-16T09:19:39.885Z',
    deletedAt: null,
    Category: { name: 'TECHNOLOGY' },
    User: {
      username: 'MCFrank16',
      firstName: 'Frank',
      lastName: 'Mutabazi',
      profileImage: 'https://res.cloudinary.com/al-tech/image/upload/v1565805494/nenxg6yetjfgbk96iq8l.jpg'
    }
  }
};

const tagPayload = {
  tags: {
    status: 200,
    data: [
      {
        id: 4,
        name: 'Software',
        createdAt: '2019-08-14T12:56:55.150Z',
        updatedAt: '2019-08-14T12:56:55.150Z'
      },
      {
        id: 6,
        name: 'Growing',
        createdAt: '2019-08-14T12:56:55.152Z',
        updatedAt: '2019-08-14T12:56:55.152Z'
      },
      {
        id: 5,
        name: 'Engineering',
        createdAt: '2019-08-14T12:56:55.151Z',
        updatedAt: '2019-08-14T12:56:55.151Z'
      }
    ]
  }
};

const errorPayload = { error: { Article: 'Article not found' } };

mapStateToProps({ readArticle: {}, populars: {} });
const match = { params: { slug: 'slug' } };

const props = {
  readArticle: jest.fn(),
  getTags: jest.fn(),
  match,
  tags: { data: [{ id: 1, name: 'TECH' }] },
  article: { Category: { name: 'Tech' }, User: { firstName: 'Frank' } },
  readPopularArticle: jest.fn(),
  isLoading: false,
  popularArticle: [{
    id: 1,
    name: 'TECH'
  }]
};
const readOneArticle = shallow(<ReadArticle {...props} />);
describe('read Article component', () => {
  it('renders the Display component', () => {
    expect(readOneArticle).toBeDefined();
    expect(readOneArticle).toMatchSnapshot();
  });

  it('will receive props', () => {
    readOneArticle.setState({
      article: {},
      slug: 'why-do-we-test',
      tags: [],
      articles: [{
        id: 4,
        name: 'Software',
        createdAt: '2019-08-14T12:56:55.150Z',
        updatedAt: '2019-08-14T12:56:55.150Z'
      }],
      isLoading: false
    });
    readOneArticle.setProps(props);
    expect(readOneArticle.instance().props).toEqual(props);
    expect(readOneArticle.instance().state.isLoading).toEqual(true);
  });
});

describe('test the reducers', () => {
  it('should handle the read article action result', () => {
    expect(Readreducer(initialState, {
      type: 'FETCH_ONE_ARTICLE',
      payload
    })).toEqual({
      Article: {},
      article: { ...payload }
    });
  });

  it('should handle the get tags by article action result', () => {
    expect(Readreducer(initialState, {
      type: 'GET_ARTICLE_TAGS',
      payload: tagPayload
    })).toEqual({
      Article: {},
      tags: { ...tagPayload }
    });
  });

  it('should handle the catch error feature', () => {
    expect(Readreducer(initialState, {
      type: 'READ_ARTICLE_ERROR',
      payload: errorPayload
    })).toEqual({
      Article: {},
      error: { ...errorPayload }
    });
  });
});

describe('testing the read article actions', () => {
  it('should dispatch the READ_ARTICLE action and payload', () => {
    mockAxios.get = jest.fn(() => Promise.resolve(payload));
    store.dispatch(readArticle(payload));
    expect(store.getActions()[0]).toMatchSnapshot();
  });

  it('should dispatch the READ_ARTICLE action and payload and throws an error', () => {
    const dataTest = { body: 'this is awesome' };
    mockAxios.get = jest.fn(() => Promise.reject({ response: { data: { errors: { Article: 'Article is not found' } } } }));
    store.dispatch(readArticle(dataTest));
    expect(store.getActions()).toMatchSnapshot();
  });


  it('should dispatch the GET_TAGS action and payload', () => {
    const dataTest = { body: 'this is awesome' };
    mockAxios.get = jest.fn(() => Promise.resolve(tagPayload));
    store.dispatch(getTags(dataTest));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should dispatch the GET_TAGS action and payload and throws an error', () => {
    const dataTest = { body: 'this is awesome' };
    mockAxios.get = jest.fn(() => Promise.reject({ response: { data: { errors: { tags: 'No tags found for this article' } } } }));
    store.dispatch(getTags(dataTest));
    expect(store.getActions()).toMatchSnapshot();
  });
});
