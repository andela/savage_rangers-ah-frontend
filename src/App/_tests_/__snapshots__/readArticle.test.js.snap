// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read Article component renders the Display component 1`] = `ShallowWrapper {}`;

exports[`testing the read article actions should dispatch the GET_TAGS action and payload 1`] = `
Array [
  Object {
    "payload": "Article is not found",
    "type": "READ_ARTICLE_ERROR",
  },
]
`;

exports[`testing the read article actions should dispatch the GET_TAGS action and payload and throws an error 1`] = `
Array [
  Object {
    "payload": "Article is not found",
    "type": "READ_ARTICLE_ERROR",
  },
  Object {
    "payload": undefined,
    "type": "GET_ARTICLE_TAGS",
  },
]
`;

exports[`testing the read article actions should dispatch the READ_ARTICLE action and payload 1`] = `undefined`;

exports[`testing the read article actions should dispatch the READ_ARTICLE action and payload and throws an error 1`] = `Array []`;
